### PR TITLE
[Merged by Bors] - Restructure lint lists in `boa_engine`

### PR DESCRIPTION
**Pull request by raskad**
This Pull Request restructures the lint deny/warn/allow lists in `boa_engine`. It adds a lot of documentation to pup functions. There are still a few clippy lints that are not fixed, mainly regarding casting of number types. Fixing those lints effectiveley would in some cases probably require bigger refactors.

This should probably wait for #2449 to be merged, because that PR already fixes that lints regarding the `Date` built-in.

### CODE REVIEW

File: boa_engine/src/builtins/array/mod.rs [hunk 1]
```diff
@@ -236,10 +236,10 @@ impl Array {
         // 3. Let A be ! MakeBasicObject(« [[Prototype]], [[Extensible]] »).
         // 4. Set A.[[Prototype]] to proto.
         // 5. Set A.[[DefineOwnProperty]] as specified in 10.4.2.1.
-        let prototype = match prototype {
-            Some(prototype) => prototype,
-            None => context.intrinsics().constructors().array().prototype(),
-        };
+        let prototype = prototype.map_or_else(
```
Comment (Razican) [inline boa_engine/src/builtins/array/mod.rs:239]: Wouldn't `unwrap_or_else()` be better suited here?

File: boa_engine/src/builtins/date/mod.rs [hunk 2]
```diff
@@ -1,3 +1,12 @@
+//! This module implements the global `Date` object.
```
Comment (Razican) [inline boa_engine/src/builtins/date/mod.rs:1]: Changes in this module should probably wait for #2449

File: boa_engine/src/builtins/intl/mod.rs [hunk 3]
```diff
@@ -281,10 +282,7 @@ fn unicode_extension_components(extension: &str) -> UniExtRecord {
         let e = extension[k..].find('-');
 
         // b. If e = -1, let len be size - k; else let len be e - k.
-        let len = match e {
-            Some(pos) => pos,
-            None => size - k,
-        };
+        let len = e.map_or(size - k, |pos| pos);
```
Comment (Razican) [inline boa_engine/src/builtins/intl/mod.rs:285]: This can probably be `unwrap_or()`, right?

**Review by Razican — COMMENTED** [review]
Great work! Only halfway there with the review, and it looks promising :)

**Review by HalidOdat — APPROVED** [review]
Looks good to me! :)

File: boa_engine/src/environments/runtime.rs [hunk 4]
```diff
@@ -408,11 +406,7 @@ impl DeclarativeEnvironmentStack {
             ThisBindingStatus::Uninitialized
         };
 
-        let this = if let Some(this) = this {
-            this
-        } else {
-            JsValue::Null
-        };
+        let this = this.map_or(JsValue::Null, |this| this);
```
Comment (Razican) [inline boa_engine/src/environments/runtime.rs:409]: Here we should use `unwrap_or()`, or even `unwrap_or_default()`

File: boa_engine/src/object/property_map.rs [hunk 5]
```diff
@@ -221,20 +221,30 @@ impl IndexedProperties {
...
     string_properties: OrderedHashMap<JsString>,
-    /// Symbol Properties
+
+    /// Properties stored with `Symbol`s a keys.
     symbol_properties: OrderedHashMap<JsSymbol>,
 }
 
 impl PropertyMap {
+    /// Create a new [`PropertyMap`].
+    #[must_use]
```
Comment (Razican) [inline boa_engine/src/object/property_map.rs:241]: I would add an inline hint here

File: boa_engine/src/value/operations.rs [hunk 6]
```diff
@@ -449,14 +461,14 @@ impl JsValue {
...
     #[inline]
     pub fn neg(&self, context: &mut Context) -> JsResult<Self> {
         Ok(match *self {
             Self::Symbol(_) | Self::Undefined => Self::new(f64::NAN),
-            Self::Object(_) => Self::new(match self.to_numeric_number(context) {
-                Ok(num) => -num,
-                Err(_) => f64::NAN,
-            }),
+            Self::Object(_) => {
+                Self::new(self.to_numeric_number(context).map_or(f64::NAN, |num| -num))
```
Comment (Razican) [inline boa_engine/src/value/operations.rs:470]: Here I would use `f64::neg` one the second part, which should more clearly indicate intent

**Review by Razican — APPROVED** [review]
I added a couple of suggestions, but I'm happy to merge it as-is :)

**Pull request closed without merging**